### PR TITLE
[SERVICE-847] Added support for 'view' entities

### PR DESCRIPTION
**Pull request by pjbroadbent**
View-based applications, including applications using Layouts v2, can now connect to the service and make API calls.

### CODE REVIEW

File: src/provider/model/FinEnvironment.ts [hunk 1]
```diff
@@ -262,4 +265,16 @@ export class FinEnvironment extends AsyncInit implements Environment {
             }
         }
     }
+
+    /**
+     * Checks if an entity is of a "page-based" entity type. These are entities that are built using a
+     * webpage/browser/DOM/etc. This only includes entity types that both meet this definition, AND are supported by
+     * the service.
+     *
+     * These entitity types have a more finely controlled handshake process, as the fin API provides better eventing
```
Comment (bryangaleopenfin) [inline src/provider/model/FinEnvironment.ts]: "entitity

File: src/provider/model/FinEnvironment.ts [hunk 2]
```diff
@@ -262,4 +265,16 @@ export class FinEnvironment extends AsyncInit implements Environment {
...
+     * and querying of these entity types.
+     */
+    private isPagedEntity(entityType: EntityType): boolean {
```
Comment (bryangaleopenfin) [inline src/provider/model/FinEnvironment.ts]: Make static or just pull out into a non-class function

**Review by bryangaleopenfin — COMMENTED** [review]
Looks good, just need to stare at `expect.unittest.ts` a bit longer

File: test/provider/expect.unittest.ts [hunk 3]
```diff
@@ -308,6 +308,16 @@ describe('When creating an external connection', () => {
 function expectTest(testWindow: TestWindow, appDirectoryResultTime: number, resultParams: ResultParam[]): void {
     const testParams = buildTestParams(testWindow, resultParams);
 
+    function getEntityType(entityIdentity: Identity, testWindowIdentity: Identity): EntityType {
```
Comment (bryangaleopenfin) [inline test/provider/expect.unittest.ts]: This feels a bit round-a-bout

Would do:

- Put `const testWindowEntityType = testWindow.appType === 'non-directory-external' ?  EntityType.EXTERNAL_CONNECTION : return EntityType.WINDOW` around line 324.

- Use that const in the `emit`s

- Put the extracted function back in-line, and use that const

**Review by bryangaleopenfin — COMMENTED** [review]
Any way to integration test this?

**Comment by bryangaleopenfin** [comment]
Back to you @pjbroadbent

**Comment by pjbroadbent** [comment]
> Any way to integration test this?

Requires creating windows with views, which would need some updates to the window creation and spawn utils. Worth considering once layouts 2 is more stable, but may just stick with some unit testing for the time being.

File: src/provider/model/FinEnvironment.ts [hunk 4]
```diff
@@ -28,6 +28,18 @@ interface KnownEntity {
...
 
+/**
+ * Checks if an entity is of a "page-based" entity type. These are entities that are built using a
+ * webpage/browser/DOM/etc. This only includes entity types that both meet this definition, AND are supported by
+ * the service.
+ *
+ * These entity types have a more finely controlled handshake process, as the fin API provides better eventing
+ * and querying of these entity types.
+ */
+function isPagedEntity(entityType: EntityType): boolean {
```
Comment (bryangaleopenfin) [inline src/provider/model/FinEnvironment.ts]: Think this more naturally belongs at the bottom, being an implementation detail of the class

File: src/provider/model/Environment.ts [hunk 5]
```diff
@@ -7,34 +7,61 @@ import {AppConnection} from './AppConnection';
...
 
     /**
-     * Indicates that a window has been closed.
+     * Indicates that an existing window-like entity has been closed.
      *
-     * Arguments: (identity: Identity)
+     * Arguments: (identity: Identity, entityType: EntityType)
      */
-    onWindowClosed: Signal<[Identity]>;
+    onWindowClosed: Signal<[Identity, EntityType]>;
```
Comment (ottah) [inline src/provider/model/Environment.ts:59]: Do we want to start referring to windows as entities now? Rename to onEntity*?

Comment (pjbroadbent) [reply src/provider/model/Environment.ts:59]: This only fires for window-like entities, so I think `onEntityClosed` would be misleading. It should really be something along the lines of `onWindowOrViewClosed`.

Discussed this briefly with Bryan, our view was that in the interests of conciseness, views should just be thought of as windows.

File: test/provider/expect.unittest.ts [hunk 6]
```diff
@@ -374,10 +401,28 @@ function expectTest(testWindow: TestWindow, appDirectoryResultTime: number, resu
...
+                };
+                mockEnvironment.onWindowCreated.emit(windowIdentity, EntityType.WINDOW);
+            }
             mockEnvironment.onWindowCreated.emit(identity, testWindowEntityType);
         }, testWindow.createdTime);
         maybeSetTimeout(() => mockApiHandler.onConnection.emit(identity), testWindow.connectionTime);
-        maybeSetTimeout(() => mockEnvironment.onWindowClosed.emit(identity, testWindowEntityType), testWindow.closeTime);
+        maybeSetTimeout(() => {
+            if (testWindow.windowType === 'view') {
+                // There will also be an event for the window creation at around the same time as the view creation
```
Comment (bryangaleopenfin) [inline test/provider/expect.unittest.ts]: Change comment for closed event

**Review by bryangaleopenfin — COMMENTED** [review]
Looks good, just the comment comment

**Review by bryangaleopenfin — APPROVED** [review]

**Review by tomrobertsOF — APPROVED** [review]

**Pull request merged**